spotlight-client/src/datastores/Metric.ts

```diff
@@ -0,0 +1,321 @@
+}
+
+type MetricsContent = NamedEntity & { methodology: string };
+
+type MetricFactoryOptionsBase = MetricsContent;
+
+type TotalIdentifier = "ALL";
+const TOTAL_KEY: TotalIdentifier = "ALL";
+
+type DemographicView = "total" | "race" | "gender" | "age";
+
+const DIMENSION_DATA_KEYS: Record<
+  Exclude<DemographicView, "total">,
+  keyof DemographicFields
+> = {
+  age: "ageBucket",
+  gender: "gender",
+  race: "raceOrEthnicity",
+};
+
+// TODO: should these be enums?
+type RaceIdentifier =
+  | TotalIdentifier
+  | "AMERICAN_INDIAN_ALASKAN_NATIVE"
+  | "ASIAN"
+  | "BLACK"
+  | "HISPANIC"
+  | "NATIVE_HAWAIIAN_PACIFIC_ISLANDER"
+  | "WHITE"
+  | "OTHER";
+type GenderIdentifier = TotalIdentifier | "FEMALE" | "MALE";
+type AgeIdentifier =
+  | TotalIdentifier
+  | "<25"
+  | "25-29"
+  | "30-34"
+  | "35-39"
+  | "40<";
+
+type BaseMetricFields = {
+  stateCode: string;
+};
+
+type DemographicFields = {
+  raceOrEthnicity: RaceIdentifier;
+  gender: GenderIdentifier;
+  ageBucket: AgeIdentifier;
+};
+
+type LocalityFields = {
+  locality: string;
+};
+
+type RawRecord = Record<string, unknown>;
+
+export type AnyMetric = Metric<RawRecord>;
+
+type PopulationBreakdownByLocationRecord = BaseMetricFields &
+  DemographicFields &
+  LocalityFields & {
+    population: number;
+  };
+
+// TODO: should these even be different types? why not just PopulationSnapshot?
+type PopulationSnapshot = Metric<PopulationBreakdownByLocationRecord>;
+
+type SentenceTypeByLocationRecord = BaseMetricFields &
+  DemographicFields &
+  LocalityFields & {
+    dualSentenceCount: number;
+    incarcerationCount: number;
+    probationCount: number;
+  };
+type SentenceTypesCurrent = Metric<SentenceTypeByLocationRecord>;
+
+type DataTransformer<RecordFormat> = (rawData: DataFile) => RecordFormat[];
+
+type InitParams<RecordFormat> = {
+  contentSource: MetricsContent;
+  dataTransformer: DataTransformer<RecordFormat>;
+  sourceFileName: string;
+  // NOTE: these fields can only be populated if the RecordFormat contains the requisite fields;
+  // however, they are not REQUIRED to be populated in those cases (e.g., population snapshots
+  // contain demographic fields but do not need to support demographic filters).
+  // I think this means it's possible to inadvertently break a filter by setting it to `undefined`
+  // without triggering a compiler error. Maybe that can be fixed but I have not done so here
+  defaultLocalityId?: RecordFormat extends LocalityFields ? string : undefined;
+  defaultDemographicView?: RecordFormat extends DemographicFields
+    ? DemographicView
+    : undefined;
+};
+
+class Metric<RecordFormat extends RawRecord> {
+  // metadata properties
+  readonly description: string;
+
+  readonly methodology: string;
+
+  readonly name: string;
+
+  // data properties
+  private dataTransformer: DataTransformer<RecordFormat>;
+
+  private sourceFileName: string;
+
+  isLoading: boolean;
+
+  private allRecords?: RecordFormat[];
```

Comment:
it's optional because it doesn't exist until the data is fetched. Presumably you could also "successfully" fetch nothing and wind up with an empty array, and you'd be able to tell those two cases apart. I don't really know how useful that will be, as evidenced by the fact that I have already glossed over it in `get records()` and substituted an empty array! 

I could imagine doing something like throwing an error instead of blithely creating an empty array, when you are trying to access data that doesn't yet exist ... it's debatable whether that would actually be a good idea

Reply:
I see, okay. I could see why this would be useful in a more dynamic runtime environment like this.